_posts/2023-07-07-python-udf.md

```diff
@@ -0,0 +1,311 @@
+# We can simply call the function through SQL, and even use the function return to eliminate the countries that never won a world cup
+con.sql("SELECT country, wc_titles(country) as world_cups from countries").fetchall()
+# [('Brazil', 5), ('Germany', 4), ('Italy', 4), ('Argentina', 2), ('Uruguay', 2), ('France', 2), ('England', 1), ('Spain', 1), ('Netherlands', None)]
+
+```
+
+### Generating Fake Data with Faker (Built-In Type UDF)
+Here is an example that demonstrates the usage of the [Faker library](https://faker.readthedocs.io/en/master/)  to generate a scalar function in DuckDB, which returns randomly generated dates. The function, named `random_date`, does not require any inputs and outputs a `DATE` column. Since Faker utilizes built-in Python types, the function directly returns them.
+One important thing to notice, is that a function that is not deterministic based on its input must be marked as having `side_effects`.
+
+```python
+import duckdb
+
+# By importing duckdb.typing we can specify DuckDB Types directly without using strings
+from duckdb.typing import *
+
+from faker import Faker
+
+# Our Python UDF generates a random date every time it's called
+def random_date():
+     fake = Faker()
+     return fake.date_between()
+```
+
+We then have to register the Python function in DuckDB using `create_function`. Since our function doesn't require any inputs, we can pass an empty list as the `argument_type_list`. As the function returns a date, we specify `DATE` from `duckdb.typing` as the `return_type`. Note that since our `random_date()` function returns a built-in Python type (`datetime.date`), we don't need to specify the UDF type.
+
+```py
+# To exemplify the effect of side-effect, let's first run the function without marking it.
+duckdb.create_function('random_date', random_date, [], DATE)
+
+# After registration, we can use the function directly via SQL
+res = duckdb.sql('select random_date() from range (3)').fetchall()
+# [(datetime.date(2003, 8, 3),), (datetime.date(2003, 8, 3),), (datetime.date(2003, 8, 3),)]
+
+# Now let's re-add the function with side-effects marked as true.
+duckdb.remove_function('random_date')
+duckdb.create_function('random_date', random_date, [], DATE, side_effects=True)
+res = duckdb.sql('select random_date() from range (3)').fetchall()
+# [(datetime.date(2020, 11, 29),), (datetime.date(2009, 5, 18),), (datetime.date(2018, 5, 24),)]
+```
+
+### Swap String Case (PyArrow Type UDF)
+One issue with using built-in types is that you don't benefit from zero-copy, vectorization and cache locality. Using PyArrow as a UDF type should be favored to leverage these optimizations.
+
+To demonstrate a PyArrow function, let's consider a simple example where we want to transform lowercase characters to uppercase and uppercase characters to lowercase. Fortunately, PyArrow already has a function for this in the compute engine, and it's as simple as calling pc.utf8_swapcase(x)
```

Comment:
`pc.utf8...` should be formatted as code and there should be a full stop at the end of the sentence.